### PR TITLE
Only find aws hosts that are running

### PR DESCRIPTION
**Pull request by jsravn**
When finding instance by node name in AWS, only retrieve running
instances when using tags to identify nodes.  Otherwise terminated, old
nodes can show up with the same tag when rebuilding nodes in the
cluster.

This was extracted from #20207

### CODE REVIEW

**Comment by balooo** [comment]
I can confirm that I authored these commits.

**Comment by justinsb** [comment]
OK thanks for extracting - this makes it much clearer.

Two things we should figure out:

1. We have a function isAlive that we should probably remove and merge into findInstanceByNodeName.  Looking through the code, we don't care about retrieving non-alive instances.  This will make the code a little more efficient also!
1. I think that getInstancesByNodeNames should behave the same way as the singular form (although if this proves difficult, we can remove getInstancesByNodeNames in favor of getAllInstances; it's only used in the ELB code when we're actually getting almost all instances)

**Comment by chbatey** [comment]
We've changed getInstancesByNodeNames to filter via the AWS API and rebased

**Comment by justinsb** [comment]
Marking for 1.2 per request of jsravn in slack; not guaranteed to make 1.2 but has been highlighted as a serious issue.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -2274,35 +2250,35 @@ func (a *AWSCloud) getInstancesByIDs(instanceIDs []*string) (map[string]*ec2.Ins
 // This is currently implemented by fetching all the instances, because this is currently called for all nodes (i.e. the majority)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Can we update this comment please to reflect the more efficient implementation?

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -2274,35 +2250,35 @@ func (a *AWSCloud) getInstancesByIDs(instanceIDs []*string) (map[string]*ec2.Ins
...
 // In practice, the breakeven vs looping through and calling getInstanceByNodeName is probably around N=2.
 func (a *AWSCloud) getInstancesByNodeNames(nodeNames []string) ([]*ec2.Instance, error) {
-	allInstances, err := a.getAllInstances()
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Is getAllInstances now unused?  If so, we can delete it.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 3]
```diff
@@ -2274,35 +2250,35 @@ func (a *AWSCloud) getInstancesByIDs(instanceIDs []*string) (map[string]*ec2.Ins
...
-	if err != nil {
-		return nil, err
+	names := make([]*string, len(nodeNames))
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: I think aws.StringSlice does this for you

**Comment by justinsb** [comment]
A few nits, and a rebase needed (totally my fault - I solved one small instance of this problem & broke your more complete fix).  But then I will LGTM.  We'll also need the CLA-bot to be happy.

**Comment by justinsb** [comment]
LGTM - thanks!

**Pull request merged**